### DISCUSSION TITLE
PNG optimization for raster tile images

### DISCUSSION THREAD

**Discussion by Camerash** [protomaps/PMTiles#295] [Q&A] [answered]
Hello!

I am trying to find ways in compressing PNG tile images inside a `.pmtiles` file.

The only way I am getting close is by manipulating the source `.mbtiles` file by a ancient (and deprecated tool) called `mbpipe` provided by Mapbox.

This tool allows iterate over the `mbtiles` file for each tile images, pipe them into an image optimization library, then replace it in place inside the `mbtiles`:
`mbpipe 'pngquant -' output.mbtiles`
Related discussion can be found [here](https://gis.stackexchange.com/questions/89068/how-to-install-and-use-mbpipe)

I then convert the `mbtiles` to `pmtiles` by:
`pmtiles convert output.mbtiles output.pmtiles`

However after the compression process, some tiles are not loading / showing properly, and the tiles that are not showing varies by each different compression attempts.
In comparison, the non-compressed `pmtiles` file work flawlessly.

I suspect it's due to how the tool `mbpipe` is ill-written and deprecated.

I think there is much value to be had in such compression, as my file went from ~400MB down to ~100MB.

Is there any method available to compress images inside `pmtiles`, or `mbtiles` files?

Thank you so much.

**Comment by bdon** [comment, marked as answer]
Renamed title compression -> optimization

How are you generating the MBTiles file? I would recommend doing the `pngquant` optimization at the source step before putting it into the tile archive. the PMTiles tooling are generally agnostic to the contents of the individual tiles. 

In any case, optimizing PNG images inside a PMTiles archive would require re-writing the entire archive, because the lengths of the tiles would change.

**Reply by Camerash** [reply]
Hello, thanks for the quick answer!

Currently I am using the [mapbox rio mbtiles](https://github.com/mapbox/rio-mbtiles) to generate the `mbtiles` file from a GeoTIFF

Seems the command itself doesn't provide a way to perform PNG optimisation step.

**Reply by Camerash** [reply]
After a bit of digging I found that mapbox have a much better supported tool called [mb-util](https://github.com/mapbox/mbutil) which are able to unpack and repack an mbtiles file.

I think I will try to unpack -> optimize -> repack and see how it goes. Thanks!

**Reply by Camerash** [reply]
I have successfully optimize pngs inside an `mbtiles` files with great success by leveraging the much more maintained tools `mb-util`. I will share my solution below.

```bash
# Extract tiles to tmp folder
mb-util out-raw.mbtiles tmp

# Optimize pngs
find tmp/ -name '*.png' -print0 | xargs -0 -P8 -L1 pngquant

# Build mbtiles again
mb-util tmp out.mbtiles

# Convert to pmtiles
pmtiles convert out.mbtiles out.pmtiles
```
